### PR TITLE
Allow cancelling selections in restart dialog

### PR DESCRIPTION
**Pull request by haanhvu**
This PR adds a back button in restart dialog. If users click it or dismiss the dialog, the selections that users made (that leaded to restart requirement) will be cancelled.

This PR was originally created to fix https://github.com/Igalia/wolvic/issues/1546 But the scope of this PR covers more than MSAA selection cancelling. It covers all the settings that require restart:
- Display options: MSAA, Display Density, Display Dpi, Reset
- Developer options: UI Hardware Acceleration, Debug Logging, WebGL Out Of Process, Reset
- Privacy options: System Root CA, Reset

### CODE REVIEW

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/dialogs/RestartDialogWidget.java [hunk 1]
```diff
@@ -12,6 +12,8 @@
 
 public class RestartDialogWidget extends PromptDialogWidget {
 
+    public CancelFunction CancelFunction;
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/ui/widgets/dialogs/RestartDialogWidget.java:15]: Nit: variables should start with lowercase characters

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/dialogs/RestartDialogWidget.java [hunk 2]
```diff
@@ -41,4 +48,8 @@ public void updateUI() {
         setTitle(R.string.restart_dialog_restart);
         setBody(getContext().getString(R.string.restart_dialog_text, getContext().getString(R.string.app_name)));
     }
+
+    public interface CancelFunction {
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/ui/widgets/dialogs/RestartDialogWidget.java:52]: I'd call it `CancelCallback` instead, it's more consistent with the naming

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/settings/DeveloperOptionsView.java [hunk 3]
```diff
@@ -134,17 +141,38 @@ protected void updateUI() {
...
             setLocalAddon(SettingsStore.LOCAL_ADDON_ALLOWED, true);
         }
 
         if (restart) {
-            showRestartDialog();
+            showRestartDialog(cancelReset(restartDebugLogging, restartUIHardwareAcceleration, restartWebGLOutOfProcess));
         }
     };
 
+    private RestartDialogWidget.CancelFunction cancelReset(boolean restartDebugLogging,
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/ui/widgets/settings/DeveloperOptionsView.java:156]: The problem with this approach is that it isn't very extensible. For each new setting we should have to add a new parameter to the method. I think we could get rid of the `restart` variable and just call directly the showRestartDialog method passing a different cancel callback

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/settings/DeveloperOptionsView.java [hunk 4]
```diff
@@ -165,10 +193,18 @@ private void setUIHardwareAcceleration(boolean value, boolean doApply) {
 
         if (doApply) {
             SettingsStore.getInstance(getContext()).setUIHardwareAccelerationEnabled(value);
-            showRestartDialog();
+            showRestartDialog(cancelUIHardwareAccelerationSelection(value));
         }
     }
 
+    private RestartDialogWidget.CancelFunction cancelUIHardwareAccelerationSelection(boolean value) {
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/ui/widgets/settings/DeveloperOptionsView.java:200]: No need to define the function, you can pass a lambda to showRestartDialog(). The same for the others

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/settings/DisplayOptionsView.java [hunk 5]
```diff
@@ -412,14 +440,22 @@ private boolean setDisplayDensity(float newDensity) {
...
         mBinding.dpiEdit.setOnClickListener(null);
         boolean restart = false;
-        int prevDensity = SettingsStore.getInstance(getContext()).getDisplayDpi();
+        int prevDpi = SettingsStore.getInstance(getContext()).getDisplayDpi();
         if (newDpi < SettingsStore.DISPLAY_DPI_MIN || newDpi > SettingsStore.DISPLAY_DPI_MAX) {
-            newDpi = prevDensity;
+            newDpi = prevDpi;
 
-        } else if (prevDensity != newDpi) {
+        } else if (prevDpi != newDpi) {
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/ui/widgets/settings/DisplayOptionsView.java:458]: This rename is interesting by itself. It'd be nice if you could split it in a different commit in this PR

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/settings/SettingsWidget.java [hunk 6]
```diff
@@ -62,6 +62,7 @@
 import mozilla.components.concept.sync.Profile;
 
 public class SettingsWidget extends UIDialog implements SettingsView.Delegate {
+    public RestartDialogWidget.CancelFunction CancelFunction;
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/ui/widgets/settings/SettingsWidget.java:65]: Ditto lowercase characters on start.

**Review by svillar — CHANGES_REQUESTED** [review]
Good job!

I've a few comments though

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/settings/DeveloperOptionsView.java (hunk 3 above)
Comment (haanhvu) [reply app/src/common/shared/com/igalia/wolvic/ui/widgets/settings/DeveloperOptionsView.java]: I made it more extensible. Please check if the latest change is what you had in mind.

**Review by felipeerias — CHANGES_REQUESTED** [review]
Thank you so much. I have added a few small notes.

There is also one larger change that I would like to discuss. I think that it would make sense to reorganize a bit the functionality so we always pass the callback when we display the restart dialog.

The basic API for this functionality would look like this:

```java
public abstract class SettingsView extends FrameLayout {

    public interface CancelCallback {
        void cancel();
    }

    public interface Delegate {
        ...
        void showRestartDialog(CancelCallback cancelCallback);
        ...
    }

    protected void showRestartDialog(SettingsView.CancelCallback cancelCallback) {
        if (mDelegate != null) {
            mDelegate.showRestartDialog(cancelCallback);
        }
    }
}
```

This API is implemented by `SettingsWidget`:

```java
public class SettingsWidget extends UIDialog implements SettingsView.Delegate {

    @Override
    public void showRestartDialog(SettingsView.CancelCallback cancelCallback) {
        if (mRestartDialog == null) {
            mRestartDialog = new RestartDialogWidget(getContext());
        }

        mRestartDialog.setCancelCallback(cancelCallback);
        mRestartDialog.show(REQUEST_FOCUS);
    }
}
```

Note that `SettingsWidget` does not need to hold on to the callback.

Also, it is better if the dialog receives the callback with a setter method.

```java
public class RestartDialogWidget extends PromptDialogWidget {

    private SettingsView.CancelCallback mCancelCallback;

    public void setCancelCallback(SettingsView.CancelCallback cancelCallback) {
        mCancelCallback = cancelCallback;
    }
}
```

What do you think? Does this make sense?

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/settings/DisplayOptionsView.java [hunk 7]
```diff
@@ -238,11 +250,16 @@ public boolean isEditing() {
         if (!mBinding.uaRadio.getValueForId(mBinding.uaRadio.getCheckedRadioButtonId()).equals(SettingsStore.UA_MODE_DEFAULT)) {
             setUaMode(mBinding.uaRadio.getIdForValue(SettingsStore.UA_MODE_DEFAULT), true);
         }
-        if (!mBinding.msaaRadio.getValueForId(mBinding.msaaRadio.getCheckedRadioButtonId()).equals(SettingsStore.MSAA_DEFAULT_LEVEL)) {
+
+        Object prevMSAA = mBinding.msaaRadio.getValueForId(mBinding.msaaRadio.getCheckedRadioButtonId());
+        if (!prevMSAA.equals(SettingsStore.MSAA_DEFAULT_LEVEL)) {
             setMSAAMode(mBinding.msaaRadio.getIdForValue(SettingsStore.MSAA_DEFAULT_LEVEL), true);
+            restart = true;
```
Comment (haanhvu) [inline app/src/common/shared/com/igalia/wolvic/ui/widgets/settings/DisplayOptionsView.java:257]: L257 is not part of this change. I added because we forgot to set restart for MSAA change in reset.

**Review by haanhvu — COMMENTED** [review]
@felipeerias I resolved all your reviews

**Review by svillar — CHANGES_REQUESTED** [review]
Much better now, but there is still a case that is not properly handling. If you dismiss the dialog by clicking outside of it then we should perform the cancel action as well.

**Comment by svillar** [comment]
I also think we should change the look and feel of that dialog, the three buttons are not properly arranged and also I think we should use some accent color for the cancel likely. @felipeerias WDYT?

**Comment by felipeerias** [comment]
@svillar Yes, the point about the buttons makes sense.

I have been trying to adjust the layout but it doesn't look great. One issue is that this dialog is used in many places, so it is hard to make big changes without side effects.

Screenshot_com igalia wolvic dev_2024 10 04-23 24 11 434_307

But I wonder if we need that third button?

Maybe it is enough to revert the user's selection if the dialog is dismissed with the Back button in the controller, or by clicking outside.

Additionally, maybe we could add a "Back" button to the dialog window like the one in the Settings? Something like this:

Screenshot_com igalia wolvic dev_2024 10 04-23 18 41 544_870

**Comment by svillar** [comment]
> @svillar Yes, the point about the buttons makes sense.
> 
> I have been trying to adjust the layout but it doesn't look great. One issue is that this dialog is used in many places, so it is hard to make big changes without side effects.
> 
> Screenshot_com igalia wolvic dev_2024 10 04-23 24 11 434_307
> 
> But I wonder if we need that third button?
> 
> Maybe it is enough to revert the user's selection if the dialog is dismissed with the Back button in the controller, or by clicking outside.
> 
> Additionally, maybe we could add a "Back" button to the dialog window like the one in the Settings? Something like this:
> 
> Screenshot_com igalia wolvic dev_2024 10 04-23 18 41 544_870

Yeah I agree with you, the cancel button does not make much sense from the UI POV. I think we both agree that we can implement the cancel action when dismissing the dialog. Optionally we could add a back button to make it more explicit.

**Comment by haanhvu** [comment]
Yeah I agree to remove the Cancel button, treat dismissing dialog as cancelling, and add a back button for cancelling. As a user I see back button as "back to my previous selection" in this case.

**Comment by felipeerias** [comment]
﻿@haanhvu I added a Back button in `prompt_dialog.xml` like this:

```xml
    <RelativeLayout
        ... 
        android:clipToPadding="false"
        ... >

        <com.igalia.wolvic.ui.views.UIButton
            android:id="@+id/backButton"
            style="?attr/navigationBarButtonStyle"
            android:layout_width="40dp"
            android:layout_marginStart="-34dp"
            android:layout_height="40dp"
            android:layout_alignParentStart="true"
            android:src="@drawable/ic_icon_back"
            android:tint="@color/midnight" />
```

It needs to be hidden by default, so we don't affect the other situations where this layout is used.

**Comment by haanhvu** [comment]
> Yeah I agree to remove the Cancel button, treat dismissing dialog as cancelling, and add a back button for cancelling. As a user I see back button as "back to my previous selection" in this case.

@felipeerias @svillar I have done all these and they worked fine with my Quest 3. Pls help review. Thanks!

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/dialogs/RestartDialogWidget.java [hunk 8]
```diff
@@ -53,6 +61,15 @@ public void updateUI() {
         setBody(getContext().getString(R.string.restart_dialog_text, getContext().getString(R.string.app_name)));
     }
 
+    @Override
+    public void onDismiss() {
+        if (noButtonsClicked && mCancelCallback != null) {
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/ui/widgets/dialogs/RestartDialogWidget.java:66]: What do you need the noButtonsClicked for?. If the dialog is dismissed we just execute the cancel callback if available and that's it. Am I missing anything?

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/settings/DeveloperOptionsView.java [hunk 9]
```diff
@@ -158,14 +166,18 @@ private void setRemoteDebugging(boolean value, boolean doApply) {
...
+
         mBinding.hardwareAccelerationSwitch.setOnCheckedChangeListener(null);
         mBinding.hardwareAccelerationSwitch.setValue(value, false);
         mBinding.hardwareAccelerationSwitch.setOnCheckedChangeListener(mUIHardwareAccelerationListener);
         mBinding.hardwareAccelerationSwitch.setVisibility(DeviceType.getType() == DeviceType.MagicLeap2 ? View.GONE : View.VISIBLE);
 
         if (doApply) {
             SettingsStore.getInstance(getContext()).setUIHardwareAccelerationEnabled(value);
-            showRestartDialog();
+            showRestartDialog(() -> {
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/ui/widgets/settings/DeveloperOptionsView.java:178]: Nit: this and the following ones can be one line

**Review by svillar — CHANGES_REQUESTED** [review]
Nice! We're closer now

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/dialogs/RestartDialogWidget.java (hunk 8 above)
Comment (haanhvu) [reply app/src/common/shared/com/igalia/wolvic/ui/widgets/dialogs/RestartDialogWidget.java:66]: Currentlly `onDismiss()` is called when:
- Users dismiss explicitly (need to cancel)
- Restart Later is clicked (no need to cancel)
- Back is clicked (cancel has already been called before onDismiss())

That's why I needed `noButtonsClicked` here, i.e. for the last two cases cancel shouldn't be called in `onDismiss()`.

But now I realized the simpler solution may be:
- Replace `onDismiss()` with `hide(REMOVE_WIDGET)` in Restart Later click
- No need to call cancel before `onDismiss()` in Back click
- `onDismiss()` is called when users dismiss explicitly or Back is clicked

This way we don't need `noButtonsClicked` anymore. You agree this is a simpler way?

Comment (svillar) [reply app/src/common/shared/com/igalia/wolvic/ui/widgets/dialogs/RestartDialogWidget.java:66]: Yes , please do so

File: app/src/main/res/values/strings.xml [hunk 10]
```diff
@@ -483,6 +483,10 @@
          from being restarted. -->
     <string name="restart_later_dialog_button">Restart Later</string>
 
+    <!-- This string represents the back button the user presses in order to cancel the selection
+         that required restart. -->
+    <string name="restart_back_dialog_button">Back</string>
```
Comment (felipeerias) [inline app/src/main/res/values/strings.xml:488]: Since this will not be shown to the user, maybe we can just use the existing `"back_button"` string so we don't need to add a new one?

**Review by felipeerias — CHANGES_REQUESTED** [review]
The functionality works great. Other than a small comment, I think that this is good to merge.

**Review by svillar — APPROVED** [review]
Cool! Let's merge it

**Comment by svillar** [comment]
This PR adds a back button in restart dialog. If users click it or dismiss the dialog, the selections that users made (that leaded to restart requirement) will be cancelled.

This PR was originally created to fix https://github.com/Igalia/wolvic/issues/1546 But the scope of this PR covers more than MSAA selection cancelling. It covers all the settings that require restart:

Display options: MSAA, Display Density, Display Dpi, Reset
Developer options: UI Hardware Acceleration, Debug Logging, WebGL Out Of Process, Reset
Privacy options: System Root CA, Reset
